File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        int result = 0;

        if (s.isEmpty()) return result;

        int idx = 0;
        int len = s.length();
        boolean isNegative = false;

        while (idx < len && s.charAt(idx) == ' ') {
            idx++;
        }
        
        if (idx == len) return result;

        if (s.charAt(idx) == '-' || s.charAt(idx) == '+') {
            isNegative = (s.charAt(idx) == '-') ? true : false;
            idx++;
        }

        while (idx < len && Character.isDigit(s.charAt(idx))) {
            int digit = s.charAt(idx) - '0';
            if (result > Integer.MAX_VALUE / 10 || (result == Integer.MAX_VALUE / 10 && digit > 7)) {
                return (isNegative) ? Integer.MIN_VALUE : Integer.MAX_VALUE;
            }

            result = result * 10 + digit;
            idx++;
        }

        return (isNegative) ? -result : result;
    }
}

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.kt
class Solution {
    fun myAtoi(s: String): Int {
        var result = 0
        
        if (s.isEmpty()) return result

        var idx = 0
        val len = s.length
        var isNegative = false

        while (idx < len && s[idx] == ' ') {
            idx++
        }

        if (idx == len) return result

        if (s[idx] == '-' || s[idx] == '+') {
            isNegative = s[idx] == '-'
            idx++
        }

        while (idx < len && s[idx].isDigit()) {
            val digit = s[idx] - '0'

            if (result > Int.MAX_VALUE / 10 || (result == Int.MAX_VALUE / 10 && digit > 7)) {
                return if (isNegative) Int.MIN_VALUE else Int.MAX_VALUE
            }

            result = result * 10 + digit
            idx++
        }

        return if (isNegative) -result else result
    }
}